project-addons/separate_purchase_orders/models/product.py

```diff
@@ -29,14 +29,18 @@ class ProductProduct(models.Model):
     _inherit = 'product.product'
 
     @api.multi
-    def _split_purchase_count(self):
+    def _purchase_count(self):
         domain = [
-            ('order_id.state', '=', 'purchase_order'),
-            ('product_id', 'in', self.mapped('id')),
+            ('product_id', 'in', self.ids),
+            ('state', 'in', ['purchase', 'done', 'purchase_order'])
         ]
-        PurchaseOrderLines = self.env['purchase.order.line'].search(domain)
+
         for product in self:
-            product.split_purchase_count = len(
-                PurchaseOrderLines.filtered(lambda r: r.product_id == product).mapped('order_id'))
+            purchase_order_lines = self.env['purchase.order.line'].search_read(domain, ['product_qty', 'state'])
```

Comment:
si usas self.ids en el domain este search_read tendría que ir antes del for y habría que tener en cuenta el producto también. 
En el código anterior había un filtered, sino se filtra, de entrar varios ids todos obtendrán el mismo valor para esos campos
Y creo que sería más eficiente usar read_group con dos agrupadores producto y estado y el parámetro lazy a False o dos read_group con distinto domain sólo agrupando por product_id